### PR TITLE
Add validation that detects repeated keys in the labels and annotations maps

### LINKED ISSUE #2965: adding the same label key twice should error
If I create an object with
labels: {
  name="hello",
  name="goodbye"
}

The object will be accepted by the apiserver, but the "hello" label will be lost.  This should either join them together, or be an error.  Silently dropping the first is bad news.

### PR DESCRIPTION
**Pull request by brendandburns**
Fixes #2965  (a nearly 2 year old feature request!)

@kubernetes/kubectl 

@eparis

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34407)

### CODE REVIEW

File: pkg/api/validation/schema.go [hunk 1]
```diff
@@ -62,6 +64,70 @@ type NullSchema struct{}
...
+			if seen[string(t)] {
+				return fmt.Errorf("duplicate key: %s", string(t))
+			} else {
+				seen[string(t)] = true
+			}
+		}
+	}
+}
+
+type ConjunctiveSchema struct {
```
Comment (smarterclayton) [inline pkg/api/validation/schema.go]: Could just declare this as `type ConjunctiveSchema []Schema`.  More idiomatic.

File: pkg/api/validation/schema.go [hunk 2]
```diff
@@ -62,6 +64,70 @@ type NullSchema struct{}
 
 func (NullSchema) ValidateBytes(data []byte) error { return nil }
 
+type NoDoubleKeySchema struct{}
+
+func (NoDoubleKeySchema) ValidateBytes(data []byte) error {
```
Comment (smarterclayton) [inline pkg/api/validation/schema.go]: Unfortunate that you have to hardcode just two fields.  This would be a lot more useful if it traversed all maps.

Comment (brendandburns) [inline pkg/api/validation/schema.go]: the trouble is that this is bytes, we could automatically build up a map for every runtime.Object I suppose.  Follow on PR?

**Comment by brendandburns** [comment]
@k8s-bot test this issue: IGNORE

(ran out of firewall quota)

File: pkg/api/validation/schema.go (hunk 1 above)
Comment (brendandburns) [inline pkg/api/validation/schema.go]: done.

**Comment by brendandburns** [comment]
Comment addressed, please take another look.

**Comment by brendandburns** [comment]
Rebased, ptal.

**Comment by brendandburns** [comment]
Friendly monday ping on this one ;)

**Comment by thockin** [comment]
I didn't realize this was assigned to me.  2 questions: 1) who knows the schema validation logic, because I don't.  2) don't we already have a JSONpath decoder somewhere?  I know we decode JSONpath for kubectl...

**Comment by smarterclayton** [comment]
I didn't think the JSONPath printer supports actual JSON (it supports types laid out with JSON tags, which is similar).

**Pull request merged**